Feature/pattern class

maxmeyer opened the pull request:
PR for #44. But this one is not completed! @ttaylorr @jasonlong Any suggestion for a good name for this data structure. What is it about? Is it kind of a input for the `generator` Is it a preset? I think about creating a Struct which holds the data and will be the input for a generator.

```
        base_color: '#933c3c',
        fill_color_dark: '#222',
        fill_color_light: '#ddd',
        stroke_color: '#000',
        stroke_opacity: 0.02,
        opacity_min: 0.02,
        opacity_max: 0.15
```

maxmeyer commented:
Is the order for adding the background and the the structure relevant? It is, isn't it? Otherwise the background will make the structure of a pattern invisible. Correct?

jasonlong commented:
First, yes the order matters since the base color is put down and the semi-transparent shapes sit on top of them.

As for that structure, that's kind of tricky because only one of those values (`base_color`) is something that can be set by the user. The rest are really constants. I guess there's no reason they couldn't be taken as params, but that's not been available thus far. If these values defined what the pattern would look like, I'd say something like "recipe" might be fitting, but these specific one are entirely about color. To use your terminology, I suppose these are "color presets".

maxmeyer commented:
ok. I named it that way. :smile:

maxmeyer commented:
I also solved "the order problem". Now it's up to the `Pattern`-class to order the svg.

maxmeyer commented:
@jasonlong 
I added some more stuff. Now there's a rake task avaiable which can be used to generate `patterns`. I use this to generate fixtures which can be used to implement end-to-end-tests. They should help to make sure that the library is really ok, after bigger changes like that one I'm preparing.

jasonlong commented:
:thumbsup: Nice, sounds good.

maxmeyer commented:
Still working on that.

maxmeyer commented:
@jasonlong Would you mind to add a travis hook?

jasonlong commented:
@maxmeyer Got Travis hooked up. I didn't add a `.travis.yml`, but feel free to add one if need be.

maxmeyer commented:
Thanks. Perfect. :+1:  This PR includes one already. Do you have any preferred ruby versions? Or do you want the support the last three release ruby versions? e.g. 2.0.0, 2.1.5, 2.2.0. I saw this kind of pattern in some other projects.

jasonlong commented:
No strong version preference. If those 2.0.0+ versions are a common pattern, that works for me.

maxmeyer commented:
Ok guys. This is the first rough version.

* [x] Api is the same from the user perspective
* [x] Tests added for most classes, even for the existing generators
* [x] Split up generator in background and structure generators
* [x] Add documentation for some of the new inspection methods - can be improved :smile:
* [x] Add rake task to generate fixtures - making it possible to generate a visual "test" - but can be used from other projects as well - see documentaiton
* [x] Add convience wrapper for testing and bootstrapping
* [x] Clean up filesystem - structure_generators, background_generators have their separated directories

@jasonlong @tarebyte @ttaylorr 
Would you mind to review this? I will make the test suite pass the travis build as fast as I can.

tarebyte commented on .travis.yaml:
```diff
@@ -0,0 +1,12 @@
+  - CI=1
+language: ruby
+bundler_args: --without debug
+rvm:
```
Are we not supporting 1.9.3?

jasonlong replied:
If we're not doing anything 2.0+ specific, it wouldn't hurt to support 1.9.3.

maxmeyer replied:
The reason I wouldn't support is this:

https://www.ruby-lang.org/en/news/2014/01/10/ruby-1-9-3-will-end-on-2015/
Support for Ruby version 1.9.3 will end on February 23, 2015

maxmeyer commented:
> No strong version preference. If those 2.0.0+ versions are a common pattern, that works for me.

It's not a 2.x-thing. It's more supporting the last three ruby branches. Another possibility would be to support all ruby-version which are not "End of Life".

maxmeyer commented:
I added a `Color`-class which makes it easier to handle conversion between different formats.

maxmeyer commented:
There was a bug in the Rakefile: `string` contained the full array, not just the "Master Markdown"-string.

maxmeyer commented:
I refactored the color class and added two generators.

jasonlong commented:
@maxmeyer I wanted to make sure that patterns that are generated with current versions of the gem and with these changes will be the same for a given string. Is that the case?

maxmeyer commented:
Ok. I checked `1.3.2` installed via rubygems and `feature/pattern_class` with the following code. The patterns are the same. The only thing which is different is the image size. Is this important, e.g. for repeating the patterns?

1. Create a directory

2. Initialize it with git

  ```
  git init
  ```

2. Create a Gemfile

  ```ruby
  source 'https://rubygems.org'
  gem 'geo_pattern', '=1.3.2'
  ```

  ```
  bundle install
  ```

3. Create a script with

  ```ruby
  #!/usr/bin/ruby

  require 'geo_pattern'
  require 'fileutils'

  $stderr.puts GeoPattern::VERSION

  string = 'Master Markdown'

  data = {
    'fixtures/generated_patterns/octagons.svg'                 => { input: string, patterns: [:octagons] },
    'fixtures/generated_patterns/overlapping_circles.svg'      => { input: string, patterns: [:overlapping_circles] },
    'fixtures/generated_patterns/plus_signs.svg'               => { input: string, patterns: [:plus_signs] },
    'fixtures/generated_patterns/xes.svg'                      => { input: string, patterns: [:xes] },
    'fixtures/generated_patterns/sine_waves.svg'               => { input: string, patterns: [:sine_waves] },
    'fixtures/generated_patterns/hexagons.svg'                 => { input: string, patterns: [:hexagons] },
    'fixtures/generated_patterns/overlapping_rings.svg'        => { input: string, patterns: [:overlapping_rings] },
    'fixtures/generated_patterns/plaid.svg'                    => { input: string, patterns: [:plaid] },
    'fixtures/generated_patterns/triangles.svg'                => { input: string, patterns: [:triangles] },
    'fixtures/generated_patterns/squares.svg'                  => { input: string, patterns: [:squares] },
    'fixtures/generated_patterns/nested_squares.svg'           => { input: string, patterns: [:nested_squares] },
    'fixtures/generated_patterns/mosaic_squares.svg'           => { input: string, patterns: [:mosaic_squares] },
    'fixtures/generated_patterns/concentric_circles.svg'       => { input: string, patterns: [:concentric_circles] },
    'fixtures/generated_patterns/diamonds.svg'                 => { input: string, patterns: [:diamonds] },
    'fixtures/generated_patterns/tessellation.svg'             => { input: string, patterns: [:tessellation] },
    'fixtures/generated_patterns/diamonds_with_color.svg'      => { input: string, patterns: [:diamonds], color: '#00ff00' },
    'fixtures/generated_patterns/diamonds_with_base_color.svg' => { input: string, patterns: [:diamonds], base_color: '#00ff00' }
  }

  data.each do |path, options|
    pattern = GeoPattern.generate(options.delete(:input), generator: options.delete(:patterns).first.to_s, **options)
    FileUtils.mkdir_p File.dirname(path)
    File.write(path, pattern.svg_string)
  end
  ```

3. Run the script

  ```
  bundle exec ./script.rb
  ```

4. Add generated files to repository

  ```
  git add .
  git ct -m "Init commit"
```

5. Change Gemfile

  ```ruby
  source 'https://rubygems.org'
  gem 'geo_pattern', git: 'https://github.com/maxmeyer/geo_pattern', branch: 'feature/pattern_class'
  ```

6. Run bundler

  ```
  bundle update
  ```

7. Run the script

  ```
  bundle exec ./script.rb
  ```

8. Diff output

   ```
   git diff
   ```

dg-ratiodata commented:
@jasonlong 

Fixed. Now all patterns look the same like before. I fixed the size problem. I also fixed a wrong string in the Rakefile for generating the fixtures: It was `Master Markdown`, but it should be `Mastering Markdown`. The fixtures are now used by the test suite. If anyone changes the library and runs the test suite, you should be quite sure, that everything is good to merge, if the tests are all green.

maxmeyer commented:
@jasonlong 

Fixed. Now all patterns look the same like before. I fixed the size problem. I also fixed a wrong string in the Rakefile for generating the fixtures: It was `Master Markdown`, but it should be `Mastering Markdown`. The fixtures are now used by the test suite. If anyone changes the library and runs the test suite, you should be quite sure, that everything is good to merge, if the tests are all green.

dg-ratiodata commented:
There's backward compatibility for the `String` implementation of the `patterns`-option, but not for the `Class`-implementation. I moved the structure generators to another directory and added a new namespace for them to reflect that change.

maxmeyer commented:
There's backward compatibility for the `String` implementation of the `patterns`-option, but not for the `Class`-implementation. I moved the structure generators to another directory and added a new namespace for them to reflect that change.

maxmeyer commented:
There's no example for the `chevrons` pattern on your github.io-profile! :warning:

jasonlong commented:
> There's no example for the chevrons pattern on your github.io-profile!

Whoa, you're right. I went ahead and added one. I can't remember what `base_color` I used for the others, but fwiw this one was generated with:

```
pattern = GeoPattern.generate("chevrons", patterns: GeoPattern::ChevronPattern, base_color: '#fc0')
```

maxmeyer commented:
Perfect. :-)

maxmeyer commented:
@jasonlong  @ttaylorr  Did you have the time to play around with the new version?

ttaylorr commented:
Not yet, sadly -- been a busy week or so.  I'll try and clone this down
this weekend and see what I think ;)

On Wed, Feb 4, 2015 at 2:23 PM, maxmeyer <notifications@github.com> wrote:

> @jasonlong <https://github.com/jasonlong> @ttaylorr
> <https://github.com/ttaylorr> Did you have the time to play around with
> the new version?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jasonlong/geo_pattern/pull/46#issuecomment-72920313>.
>



-- 
Thanks,
Taylor Blau

ttaylorr.com

maxmeyer commented:
Thanks a lot.

maxmeyer commented:
The last commits I added bring backward compatibility and tests for the "old style" class names. I also switched to "named parameters" (>= ruby 2) because I found myself using `pattern:` instead of using `patterns:` and wondering why the tests were failing. With the named parameters I would have got an "ArgumentError" at once, which had shown me, what was wrong.

maxmeyer commented:
@ttaylorr 
:smile:  Any problems?

ttaylorr commented:
Just gave this a good look, and: :shipit:

jasonlong commented:
Are we still waiting for a fix for the build failure?

maxmeyer commented:
@jasonlong I think you need to add the .travis.yml to the master-branch to make it work. You can use the one given in `feature/pattern_class`. The support rubies should be > 2, because I introduced named parameters in one of the last commits.

jasonlong commented:
Sounds good. I'll make sure the next release mentions the > 2 version requirement.

ttaylorr commented:
Woot!

On Sat, Feb 14, 2015 at 9:34 AM, Jason Long <notifications@github.com>
wrote:

> Merged #46 <https://github.com/jasonlong/geo_pattern/pull/46>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jasonlong/geo_pattern/pull/46#event-235848206>.
>



-- 
Thanks,
Taylor Blau

ttaylorr.com

maxmeyer commented:
> Sounds good. I'll make sure the next release mentions the > 2 version requirement.

Perfect. 

BTW:
This [line](https://github.com/jasonlong/geo_pattern/blob/master/geo_pattern.gemspec#L21) in the `gemspec` will prevent users from accidently installing the gem on a `ruby < 2`. :smile: They will get an error message when installing this version with a non-compatible ruby.

jasonlong commented:
:thumbsup: Version `1.4.0` has been published.

dg-ratiodata commented:
@jasonlong 

Should we make an announcement for 1.4.0, e.g. on http://rubyflow.com?

jasonlong commented:
> Should we make an announcement for 1.4.0, e.g. on http://rubyflow.com?

If you'd like to write something up, that's :ok_hand: by me.
